tests/unit/lms/models/oauth2_token_test.py

```diff
@@ -8,118 +8,77 @@
 
         with pytest.raises(
             sqlalchemy.exc.IntegrityError,
             match='null value in column "consumer_key" violates not-null constraint',
         ):
             db_session.flush()
 
-    def test_setting_consumer_key_sets_application_instance(
-        self, application_instance, db_session, init_kwargs
-    ):
-        token = OAuth2Token(**init_kwargs)
-        db_session.add(token)
-        db_session.flush()
+    def test_setting_consumer_key_sets_application_instance(self, db_session):
+        # For this test we need an OAuth2Token factory that doesn't pass an
+        # application_instance argument to OAuth2Token() (not even
+        # application_instance=None).
+        class OAuth2Token(factories.OAuth2Token):
+            class Meta:
+                exclude = ("application_instance",)
 
-        assert token.application_instance == application_instance
+        application_instance = factories.ApplicationInstance()
+
+        token = OAuth2Token(consumer_key=application_instance.consumer_key)
 
-    def test_setting_application_instance_sets_consumer_key(
-        self, application_instance, db_session, init_kwargs
-    ):
-        del init_kwargs["consumer_key"]
-        init_kwargs["application_instance"] = application_instance
-        token = OAuth2Token(**init_kwargs)
-        db_session.add(token)
         db_session.flush()
+        assert token.application_instance == application_instance
+
+    def test_setting_application_instance_sets_consumer_key(self, db_session):
+        application_instance = factories.ApplicationInstance()
+        token = factories.OAuth2Token(application_instance=application_instance)
 
+        db_session.flush()
         assert token.consumer_key == application_instance.consumer_key
 
-    def test_access_token_cant_be_None(self, db_session, init_kwargs):
-        del init_kwargs["access_token"]
-        db_session.add(OAuth2Token(**init_kwargs))
+    def test_access_token_cant_be_None(self, db_session):
+        factories.OAuth2Token(access_token=None)
 
         with pytest.raises(
             sqlalchemy.exc.IntegrityError,
             match='null value in column "access_token" violates not-null constraint',
         ):
             db_session.flush()
 
-    def test_refresh_token_defaults_to_None(self, db_session, init_kwargs):
-        token = OAuth2Token(**init_kwargs)
-        db_session.add(token)
+    def test_refresh_token_defaults_to_None(self, db_session):
+        token = factories.OAuth2Token(refresh_token=None)
 
         db_session.flush()
 
         assert token.refresh_token is None
 
-    def test_expires_in_defaults_to_None(self, db_session, init_kwargs):
-        token = OAuth2Token(**init_kwargs)
-        db_session.add(token)
+    def test_expires_in_defaults_to_None(self, db_session):
+        token = factories.OAuth2Token()
 
         db_session.flush()
 
         assert token.expires_in is None
```

Comment:
I think the way this particular set of tests work is a no-no. We should not be testing the factory created version of an object in the tests of that object. I think this particular test shows the clearest example of why.

Currently the factory does not set a value for expires in, but it absolutely could / should. It would be totally reasonable to add a randomizer to it so that it was `None` 50% of the time. This would intermittently (based on the repeatable randomness) break this test. This shouldn't be possible.

This is because this test is testing the compound of the model class _and_ the factory object. The factory does not have to behave like the object, _particularly_ in the way the defaults are setup. So is _absolutely mandatory_ it not be used for testing initialisation behavior, but I'd say in general it's better not to use it to test _any_ behavior of the object in case the way it works is different in a more subtle way.

Reply:
I've changed this back to using an `init_kwargs` fixture, but the fixture now uses a factory to generate a random dict instead of just returning a hardcoded dict

Reply:
I think that gives us a best of both situation.